### PR TITLE
Improve action log display with control chars

### LINKED ISSUE #23680: Incorrect line number display of process logs in actions
### Description

https://gitea.com/yp05327/testrepo/actions/runs/14
image

### Gitea Version

1.20.0+dev-251-g9f39def99

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

gitea.com

### Database

None

### PR DESCRIPTION
**Pull request by wxiaoguang**
Close #23680

Some CLI programs use "\r" and control chars to print new content in current line.

So, the strings in one line are actually from `\rReading...1%\rReading...5%\rReading...100%`

This PR tries to make the output better.

### CODE REVIEW

**Review by lunny — APPROVED** [review]

**Comment by wxiaoguang** [comment]
> Would this have adverse effects of Windows-style newlines `\r\n`? Would suggest adding a test for that as well.

Nope, you can check the tests, one of them covers the ending "\r".

**Comment by wxiaoguang** [comment]
To make code more strict, I make the function remove the trailing "\r\n" or "\n" in 2c72257128aa79179f0fd1b26b7e9b4232549cbf , then it doesn't matter whether the line ends with new-line or not.

File: web_src/js/components/RepoActionView.vue [hunk 1]
```diff
@@ -214,7 +214,7 @@ const sfc = {
 
       const logMessage = document.createElement('div');
       logMessage.className = 'log-msg';
-      logMessage.innerHTML = this.ansiToHTML.toHtml(line.message);
+      logMessage.innerHTML = this.ansiToHTML.toHtml(processConsoleLine(line.message));
       div.appendChild(logMessage);
```
Comment (silverwind) [inline web_src/js/components/RepoActionView.vue:218]: Would this variable be better called `logLine`?

**Review by wolfogre — APPROVED** [review]

**Comment by yp05327** [comment]
image: docker:23
image
tested in https://github.com/go-gitea/gitea/pull/23820/commits/2c72257128aa79179f0fd1b26b7e9b4232549cbf

I prefer add a special process to handle `\r` or `\n` which can exchange one line log into several new lines, like:
```
1 Reading...10%
2 Reading...30%
3 Reading...50%
4 Reading...70%
5 Reading...Done
```

**Comment by wxiaoguang** [comment]
I see, I haven't spent enough time on this problem and it looks like not that simple. Maybe the comment `TODO: control chars` is also necessary for most cases.


So let's mark this PR as WIP.

To make a further improvement, I guess there could be 3 approaches:

1. Make backend split "\r\n" and "\n" and "\r", then the frontend can render:

   ```
   1 Reading... 1%
   2 Reading... 5%
   3 Reading... Done
   4 Next line
   ```

2. Make frontend split:

   ```
   1 Reading... 1%
     Reading... 5%
     Reading... Done
   2 Next line
   ```

3. Wait for more time to support all control chars and test for various cases.

Feel free to help or propose new PRs.

**Comment by yp05327** [comment]
I agree approach 2 and 3. And using an existing package to handle it looks better, as we don’t need to maintain it by ourselves to support all possible chars and cases.
But I have no idea about it.

**Comment by wxiaoguang** [comment]
Thank you. I think 9d358903eaa0ba54344f4d5847f304b00fa13472 could implement approach 2. If it looks good to you, let's move on ~~

I have done a quick test, the "\n" looks good to me 😁

**Comment by yp05327** [comment]
image
image
tested in https://github.com/go-gitea/gitea/commit/9d358903eaa0ba54344f4d5847f304b00fa13472

I found the log file, and it seems that the reason of this problem is `ESC`.
image
image

**Comment by yp05327** [comment]
There is a document maybe related?
https://man7.org/linux/man-pages/man4/console_codes.4.html

**Comment by yp05327** [comment]
Remove them or handle them by packages?

I found some solutions maybe related.
- Remove them
https://stackoverflow.com/questions/14693701/how-can-i-remove-the-ansi-escape-sequences-from-a-string-in-python
- Handle them
https://github.com/k0kubun/go-ansi

**Comment by wxiaoguang** [comment]
That's also the comment in code `control chars like "\033["` , it should be handled by the `AnsiToHTML` packages (the color is also provided by these control chars). I will write some new tests to see why `AnsiToHTML` doesn't work for them in this case.

**Comment by wxiaoguang** [comment]
Hmm, I think it's a bug in `AnsiToHTML` package.

Without our code, `AnsiToHTML` itself doesn't work for `\033[1A` nor `\033[1B`, it only handles the `\033[2K`

```
  const ath = new AnsiToHTML({escapeXML: true});
  expect(ath.toHtml('\x1b[1A\x1b[2K\rtest\r\x1b[1B\x1b[1A\x1b[2K')).toEqual('A\rtest\rBA');
  expect(ath.toHtml('\x1b[1A\x1b[2Ktest\x1b[1B\x1b[1A\x1b[2K')).toEqual('AtestBA');
```

**Comment by wxiaoguang** [comment]
Although it's related to upstream bug/limitation (https://github.com/rburns/ansi-to-html) , upstream has many open issues/PRs, so I guess we could bypass the bug on our side first. And I guess some behaviors we like couldn't be accepted by upstream easily.

So, let's try to use 6f33b2eb0fca53a9bb3bc23cc32f60868c008c38 to patch.

**Comment by silverwind** [comment]
Maybe there better JS modules around for this. Many parts of GitHub's JS are open source, so I wonder if they may have their terminal rendering also available somewhere 😉.

**Comment by yp05327** [comment]
image
image
image
tested in https://github.com/go-gitea/gitea/pull/23820/commits/2604c226c4d43bee8d2c3c28dbeea9e775473eb3

**Comment by wxiaoguang** [comment]
I think it's much better than before.

Do you prefer to see full progress list or the only one final line?

* If you prefer to see the progress list, then current approach seems good enough.
* If you prefer to see the only final line, I think it's still feasible, we can only take the last message after `[0K`.
   Then the `docker pull ....` line 3 becomes one line : `Digest: .....`
   And the `apk add ...` line 16 becomes one line: `Executing busybox-......`

**Pull request merged**